diff --git i/CMakeLists.txt w/CMakeLists.txt
index cf1acb9..801c9fa 100644
--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -46,12 +46,8 @@ message(STATUS "Enable C++ trace: ${XGRAMMAR_ENABLE_CPPTRACE}")
 if(MSVC)
   set(CMAKE_CXX_FLAGS "/Wall ${CMAKE_CXX_FLAGS}")
 else()
-  if(NOT CMAKE_BUILD_TYPE STREQUAL "Debug")
-    set(CMAKE_CXX_FLAGS "-Ofast ${CMAKE_CXX_FLAGS}")
-  endif()
-
   set(CMAKE_CXX_FLAGS "-Wall -Wextra -Werror -Wno-pedantic -Wno-unused-parameter \
--Woverloaded-virtual -flto=auto ${CMAKE_CXX_FLAGS}"
+-Woverloaded-virtual ${CMAKE_CXX_FLAGS}"
   )
 endif()
 
diff --git i/pyproject.toml w/pyproject.toml
index f7acfcc..7601a74 100644
--- i/pyproject.toml
+++ w/pyproject.toml
@@ -20,8 +20,6 @@ dependencies = [
   "torch>=1.10.0",
   "transformers>=4.38.0",
   "triton; platform_system == 'linux' and platform_machine == 'x86_64'",
-  "ninja",
-  "nanobind>=2.0.0",
 ]
 dynamic = ["version"]
 
